psbt/extractor.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2018 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package psbt
+
+// The Extractor requires provision of a single PSBT
+// in which all necessary signatures are encoded, and
+// uses it to construct a fully valid network serialized
+// transaction.
+
+import (
+	"bytes"
+
+	"io"
+
+	"github.com/btcsuite/btcd/wire"
+)
+
+// Extractor takes a finalized psbt and outputs a network serialization
+func Extract(p *Psbt) ([]byte, error) {
+	if !p.IsComplete() {
+		return nil, ErrFinalizationFailure
+	}
+	var err error
+	// We take the existing UnsignedTx field and append SignatureScript
+	// and Witness as appropriate, then allow MsgTx to do the serialization
+	// for us.
+	newTx := p.UnsignedTx.Copy()
+	for i, tin := range newTx.TxIn {
+		kvPairList := (*p.InKVPairs)[i]
+		if kvPairList.containsKeyType(PsbtInFinalScriptSig) {
+			scriptSig, err := kvPairList.getValueFromKeyType(
+				PsbtInFinalScriptSig)
+			if err != nil {
+				return nil, err
+			}
+			tin.SignatureScript = scriptSig
+		}
+		if kvPairList.containsKeyType(PsbtInFinalScriptWitness) {
+			scriptWitness, err := kvPairList.getValueFromKeyType(
+				PsbtInFinalScriptWitness)
+			if err != nil {
+				return nil, err
+			}
+			// to set the witness, need to re-deserialize the field
+			// For each input, the witness is encoded as a stack
+			// with one or more items. Therefore, we first read a
+			// varint which encodes the number of stack items.
+			r := bytes.NewReader(scriptWitness)
+			witCount, err := wire.ReadVarInt(r, 0)
+			if err != nil {
+				return nil, err
+			}
+
+			// Then for witCount number of stack items, each item
+			// has a varint length prefix, followed by the witness
+			// item itself.
+			tin.Witness = make([][]byte, witCount)
+			for j := uint64(0); j < witCount; j++ {
+				count, err := wire.ReadVarInt(r, 0)
```

Comment:
This section can be extracted out as wire.ReadVarBytes

Reply:
Ah, thanks, good catch!

Reply:
done: 2d9dc6ae54f887d6cf29aa6c6e46eacfaa82d7de